### PR TITLE
feat: membership

### PR DESCRIPTION
**Pull request by richard-ramos**
## Description

Adds membership management as described in https://github.com/waku-org/waku-rlnv2-contract/issues/11 and https://github.com/waku-org/specs/blob/master/standards/core/rln-contract.md .

Some test units are still pending but the contract is already in a state in which comments on it can be made. 
I added some observations too about some points in the contract which require clarification.

Some changes regarding the spec were made based on comments on https://github.com/waku-org/waku-rlnv2-contract/issues/11 as well as removing the need of having an actual membership state stored, since this can be determined based on the expiration dates and existence of the membership in the contract.

To track the membership age I implemented this as a double linked list. It's worth mentioning that keeping track of the expired memberships means some gas expense due to having to maintain the linking for the memberships in this list. I'm exploring whether using array that grows infinitely as a list and keeping track of the index in which to start navigating the array, to see if I can save some gas this way. Suggestions of other data structures to keep track of this information are appreciated, although maybe we could consider not keeping track of the oldest memberships and have this indexing being done offchain.

Tomorrow I expect to finish the pending test units based on this implementation.

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [x] Added natspec comments?
- [x] Ran `pnpm adorno`?

### CODE REVIEW

File: src/IPriceCalculator.sol [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+interface IPriceCalculator {
+    /// Returns the token and price to pay in `token` for some `_rateLimit`
+    /// @param _rateLimit the rate limit the user wants to acquire
+    /// @param _numberOfPeriods the number of periods the user wants to acquire
+    /// @return address of the erc20 token
+    /// @return uint price to pay for acquiring the specified `_rateLimit`
+    function calculate(uint256 _rateLimit, uint32 _numberOfPeriods) external view returns (address, uint256);
```
Comment (richard-ramos) [inline src/IPriceCalculator.sol:10]: Besides the limit and number of periods, is there any other information we might need in the future to calculate the price? currently only these items are needed for linear price calculation

File: src/Membership.sol [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+
+    struct MembershipInfo {
+        /// @notice idCommitment of the previous membership
+        uint256 prev;
+        /// @notice idCommitment of the next membership
+        uint256 next;
+        /// @notice amount of the token used to acquire this membership
+        uint256 amount;
+        /// @notice numPeriods
+        uint32 numberOfPeriods;
```
Comment (richard-ramos) [inline src/Membership.sol:84]: What's the period length in days and maximum number of periods we expect? depending on this we can change the variable type. If we assume billing periods are 30d, then an `uint8` would allow us to specify a max of 255 months = 21y, meaning that uint8 would be large enough for the purposes. It all depends on the values we choose for `billingPeriod` and maximum number of periods

File: src/Membership.sol [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+error CantEraseMembership(uint256 idCommitment);
+
+contract Membership {
+    using SafeERC20 for IERC20;
+
+    /// @notice Address of the Price Calculator used to calculate the price of a new membership
+    IPriceCalculator public priceCalculator;
+
+    /// @notice Maximum total rate limit of all memberships in the tree
+    uint32 public maxTotalRateLimitPerEpoch;
```
Comment (richard-ramos) [inline src/Membership.sol:41]: This would let us have a maximum total rate limit per epoch of `4_294_967_295`. This should be large enough. if we change it to `uint16` we can potentially use a maximum ratelimit of `65_535` messages per epoch. Is this value large enough?

File: src/Membership.sol (hunk 2 above)
Comment (richard-ramos) [reply src/Membership.sol:84]: Note I'm storing the number of periods here, so when we extend the membership we calculate the new expiration date based on the same number of periods used originally (otherwise an user could register with a small number of period, and transfer the tokens associated to this period, and then extend for a much larger number of periods with no cost.

Comment (richard-ramos) [reply src/Membership.sol:84]: Worth mentioning too that storing the number of periods incur in extra gas cost for the user

File: src/Membership.sol [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+        /// @notice token used to acquire this membership
+        address token;
+    }
+
+    /// @notice Emitted when a membership is erased due to having exceeded the grace period or the owner having chosen
+    /// to not extend it
+    /// @param idCommitment the idCommitment of the member
+    /// @param userMessageLimit the rate limit of this membership
+    /// @param index the index of the membership in the merkle tree
+    event MemberExpired(uint256 idCommitment, uint32 userMessageLimit, uint32 index);
```
Comment (richard-ramos) [inline src/Membership.sol:104]: For nwaku, it would only require the index, right? I added the idCommitment and userMessageLimit in case we want to calculate the rateCommitment on the nwaku side, but is it really necessary?

File: src/Membership.sol [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+        billingPeriod = _expirationTerm;
+        gracePeriod = _gracePeriod;
+    }
+
+    /// @notice Checks if a user message limit is valid. This does not take into account whether we the total
+    /// memberships have reached already the `maxTotalRateLimitPerEpoch`
+    /// @param userMessageLimit The user message limit
+    /// @return true if the user message limit is valid, false otherwise
+    function isValidUserMessageLimit(uint32 userMessageLimit) external view returns (bool) {
+        return userMessageLimit >= minRateLimitPerMembership && userMessageLimit <= maxRateLimitPerMembership;
```
Comment (richard-ramos) [inline src/Membership.sol:143]: This does not take into account the expired memberships, although since it's an external function I could attempt to optionally calculate whether the rate limit of currently expired memberships would allow the userMessageLimit to be registered. WDYT?

File: src/Membership.sol [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+    function _acquireMembership(
+        address _sender,
+        uint256 _idCommitment,
+        uint32 _rateLimit,
+        uint32 _numberOfPeriods
+    )
+        internal
+        returns (uint32 index, bool reusedIndex)
+    {
+        if (_numberOfPeriods == 0) revert NumberOfPeriodsCantBeZero();
```
Comment (richard-ramos) [inline src/Membership.sol:162]: Should we set a maximum too?

File: src/Membership.sol [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+
+        (address token, uint256 amount) = priceCalculator.calculate(_rateLimit, _numberOfPeriods);
+        (index, reusedIndex) =
+            _setupMembershipDetails(_sender, _idCommitment, _rateLimit, _numberOfPeriods, token, amount);
+        _transferFees(_sender, token, amount);
+    }
+
+    function _transferFees(address _from, address _token, uint256 _amount) internal {
+        if (_token == address(0)) {
```
Comment (richard-ramos) [inline src/Membership.sol:171]: Contract accepts ETH too. For this the token to set in the price calculator is `address(0)`.

File: src/IPriceCalculator.sol [hunk 8]
```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+interface IPriceCalculator {
```
Comment (richard-ramos) [inline src/IPriceCalculator.sol:4]: Having the price calculation being done on a separate contract means that if in the future we want to move away from linear calculation to a curve, it should be possible by just changing the current `priceCalculator` used by the `Membership` contract

File: src/Membership.sol [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+    }
+
+    /// @dev Extend a membership expiration date. Membership must be on grace period
+    /// @param _sender the address of the holder of the membership
+    /// @param _idCommitment the idCommitment of the membership
+    function _extendMembership(address _sender, uint256 _idCommitment) public {
+        MembershipInfo storage mdetails = members[_idCommitment];
+
+        if (!_isGracePeriod(mdetails.gracePeriodStartDate, mdetails.gracePeriod, mdetails.numberOfPeriods)) {
+            // TODO: can a membership that has exceeded the expired period be extended?
```
Comment (richard-ramos) [inline src/Membership.sol:284]: ?

File: src/Membership.sol [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+    )
+        internal
+        returns (uint32 index, bool reusedIndex)
+    {
+        if (_rateLimit < minRateLimitPerMembership || _rateLimit > maxRateLimitPerMembership) {
+            revert InvalidRateLimit();
+        }
+
+        // Attempt to free expired membership slots
+        while (totalRateLimitPerEpoch + _rateLimit > maxTotalRateLimitPerEpoch) {
```
Comment (richard-ramos) [inline src/Membership.sol:206]: This will try to remove expired memberships sequentially, but it's possible that the accumulated rate limit of all the memberships being removed is not enough for the chosen rate limit to acquire. Should we also accept an optional list of memberships to expire, so the user of the contract can explore the contract offchain and pass a list expired memberships whose accumulated rate limits is enough to acquire the rate limit they need?

Comment (richard-ramos) [reply src/Membership.sol:206]: Something interesting here is to prefer inserting new elements on the tree rather than attempt to erase expired memberships first, mainly because there's no guarantee that an expired membership rate limit will match the rate limit to acquire. (and then having to iterate  to obtain enough rate limit from expired memberships).

File: src/Membership.sol (hunk 7 above)
Comment (s-tikhomirov) [reply src/Membership.sol:171]: Shouldn't the contract call revert if ETH is being sent to it? The spec only mentions DAI as the accepted token. Also, for accepting ETH we'd need to do price conversion from USD to ETH, for which we'd need an oracle. I'd say that for the first implementation accepting ETH should be forbidden.

File: src/Membership.sol (hunk 9 above)
Comment (s-tikhomirov) [reply src/Membership.sol:294]: I'd say no: extensions are only possible from _GracePeriod_ as per spec.

However, I'm not sure I inderstand what "exceeded the expired period" means. When the membership is in _Expired_ state, it stays there forever unless either the holder withdraws the deposit, or another membership overwrites the slot... Or what does "expired period" mean here?

File: src/Membership.sol (hunk 10 above)
Comment (s-tikhomirov) [reply src/Membership.sol]: > there's no guarantee that an expired membership rate limit will match the rate limit to acquire

Is it technically possible to track / predict how much expired rate limit capacity we will have at a given point in time? (So that we can quickly tell if it's even worth it to start iterating through expired membership in an attempt to overwrite them, or their rate limit won't suffice anyway.)

**Comment by s-tikhomirov** [comment]
Left a few comments, will look into the code in more detail next week, thank you!

One meta-point I'd like to make for now: I think we should discuss and make a decision on @alrevuelta's  [proposal](https://github.com/waku-org/specs/pull/34#discussion_r1743943629) about billing periods. If we conclude that this is a good idea, then I think we should first amend the spec to reflect these changes, and then review the code to ensure it matches the updated spec.

File: src/Membership.sol (hunk 3 above)
Comment (fryorcraken) [reply src/Membership.sol:40]: > use a maximum ratelimit of 65_535 messages per epoch. Is this value large enough?

Considering the memory usage to keep track. I believe `uint16` is large enough. See https://github.com/waku-org/specs/blob/master/standards/core/rln-contract.md#why-this-particular-epoch-length

File: src/Membership.sol (hunk 6 above)
Comment (fryorcraken) [reply src/Membership.sol:161]: Not sure I totally understand what you mean by "period" but when acquiring a membership, the only expiration time available should be 180 days. This value should be a parameter in the smart contract (`Membership expiration term`) that can be changed.

Or does period mean something else?

File: src/Membership.sol (hunk 10 above)
Comment (fryorcraken) [reply src/Membership.sol]: A proposal we discussed in a call was that:
- optional "slot to override" parameter when doing the insertion
- if no _slot to override_ specified, then it tries to grab next available slot in tree. If global rate limit gets exceeded, it fials
- if _slot to override_ is specified:
    - if _slot to override_ is not over grace period, fails
    - if _slot to override rate limit_ < _rate limit of new membership_, fails
   Otherwise, replace slot.
   
 This means that in terms of gas cost, it's more predictable:
 
- there is space, low gas cost
- no space, most specific membership to override, higher gas cost

than the previous idea of "override if there is something to override, otherwise do not".

Also, we should have a call on contract to purge memberships over grace period. so we can be the one to cover the purge cost.

File: .gas-snapshot [hunk 11]
```diff
@@ -1,14 +1,11 @@
...
+WakuRlnV2Test:test__IdCommitmentToMetadata__DoesntExist() (gas: 27478)
+WakuRlnV2Test:test__InsertionNormalOrder(uint32) (runs: 1001, μ: 1123469, ~: 494837)
+WakuRlnV2Test:test__InvalidPaginationQuery__EndIndexGTcommitmentIndex() (gas: 18261)
+WakuRlnV2Test:test__InvalidPaginationQuery__StartIndexGTEndIndex() (gas: 16108)
+WakuRlnV2Test:test__InvalidRegistration__DuplicateIdCommitment() (gas: 249370)
+WakuRlnV2Test:test__InvalidRegistration__InvalidIdCommitment__Zero() (gas: 12135)
+WakuRlnV2Test:test__Upgrade() (gas: 6728616)
+WakuRlnV2Test:test__ValidPaginationQuery(uint32) (runs: 1002, μ: 226428, ~: 52927)
+WakuRlnV2Test:test__ValidPaginationQuery__OneElement() (gas: 262351)
+WakuRlnV2Test:test__ValidRegistration(uint256,uint32) (runs: 1001, μ: 268741, ~: 268741)
```
Comment (alrevuelta) [inline .gas-snapshot:10]: a valid registration seems to go from `124408` to `268741` of gas, which is x2. Can we optimize it?

>  It's worth mentioning that keeping track of the expired memberships means some gas expense due to having to maintain the linking for the memberships in this list

Ideally the contract should be as simple as possible, and I would say that's possible but we may have to circle back to the spec. Perhaps some things can be handled offchain, or well, its state does not need to be tracked onchain.

File: src/Membership.sol (hunk 6 above)
Comment (alrevuelta) [reply src/Membership.sol:161]: @fryorcraken It's related to this https://github.com/waku-org/specs/pull/34/files#r1743943629

In short, instead of allowing to register a fixed amount of 180 days I'm suggesting to allow registering `x` times a `period`. We can lower this period to eg 30days so that its closer to web2. And this would allow to register for 30*1days, 30*2 days, 30*3days, etc. the user can choose.

This also lowers the entry price. Eg would you pay for a service that requires upfront payment of 180 days? 

This also related to an idea that is being cooked regarding incentives, where the fees of a given "period" are distributed to node operators. Will share once ready.

File: src/Membership.sol (hunk 9 above)
Comment (alrevuelta) [reply src/Membership.sol:305]: > I'd say no: extensions are only possible from GracePeriod as per spec.

any reason why imposing this limit? I would say allowing to renew/extend at any time would be desirable.

**Review by alrevuelta — COMMENTED** [review]
Left some comments. I'm a bit concerned with the x2 gas increase.

File: src/Membership.sol (hunk 9 above)
Comment (s-tikhomirov) [reply src/Membership.sol:305]: > any reason why imposing this limit? I would say allowing to renew/extend at any time would be desirable.

The spec FAQ says:

> The rationale is to make possible parameter changes that the contract _Owner_ might make (e.g., for security reasons) applicable to most memberships.

In other words, we want to limit extensions somewhat to pervent users from keeping old membership parameters for too long if the contract Owner changes them.

However, I feel that it isn't a really solid justification. Strictly speaking, there are multiple design goals that we can aim for in parallel:

1. make the state transition diagram as simple as possible;
2. make the UX familiar and friendly;
3. prevent users from avoiding parameter changes that the contract Owner might introduce.

Only allowing extensions from GracePeriod helps (1), hurts (2), and... is neutral towards (3) (though I'm not sure about the last statement).

Comment (alrevuelta) [reply src/Membership.sol:305]: > In other words, we want to limit extensions somewhat to pervent users from keeping old membership parameters for too long if the contract Owner changes them.

This makes sense, but we are talking about two different things? I mean you can allow extensions at any time (not just from grace period) but also limit it to X to prevent "old membership parameters for too long if the contract Owner changes them."

Comment (alrevuelta) [reply src/Membership.sol:305]: I mean you can allow extending at any time as long as the total remaining time is not more than x days.

File: src/Membership.sol (hunk 7 above)
Comment (richard-ramos) [reply src/Membership.sol:170]: I added the functionality to make the contract future proof, and would be up to us to determine whether we want to use ETH or not, but since there's no immediate plans for this, I'll drop that part of the code, specially since it would mean saving few gas due to not have to do the comparison against address(0). 
We can always add this back via an upgrade

File: .gas-snapshot (hunk 11 above)
Comment (richard-ramos) [reply .gas-snapshot:10]: I can try optimizing it, but the main problem IMO is the additional storage required due to https://github.com/waku-org/waku-rlnv2-contract/blob/251d8907d4d7ee71e77fdb9ae08e1d4cd32f9f23/src/Membership.sol#L39-L96

We could probably cut back the cost by not mainainting the ordered membership list, and have nwaku be the one tracking the age of the memberships and deciding then which memberships to erase. I'd actually prefer that since we make the contract cheaper for the end user, but I don't know if keeping track of the membership age in nwaku is something we want to do?

File: src/Membership.sol (hunk 10 above)
Comment (richard-ramos) [reply src/Membership.sol]: > Is it technically possible to track / predict how much expired rate limit capacity we will have at a given point in time? (So that we can quickly tell if it's even worth it to start iterating through expired membership in an attempt to overwrite them, or their rate limit won't suffice anyway.)

I'd say this can be done offchain and then via function parameters indicate to the contract whether we want to: just insert a new leaf in the tree / have the contract remove automatically expired / remove specific memberships. Then nwaku could call `estimateGas` on any of these options and choose the one that incurs in less gas cost.

> A proposal we discussed in a call ...

Ah! this seems to be aligned with what I just mentioned

> TLDR: do we need to  grab next available slot in tree every time, or is it perhaps better to do this (more) lazily?

This is hard to answer without testing.  Perhaps we can have the three options I mentioned earlier, and have some script to try to determine the behavior of the contract once we have a large enough number of memberships.

File: src/Membership.sol (hunk 6 above)
Comment (s-tikhomirov) [reply src/Membership.sol:161]: As decided on the research call today (2024-09-09), let's set aside the billing period idea for now, and review it more carefully for the next spec iteration. I've created an [issue](https://github.com/waku-org/specs/issues/37) for that purpose, we can move the discussion there.

File: src/Membership.sol (hunk 9 above)
Comment (s-tikhomirov) [reply src/Membership.sol:305]: Created an [issue](https://github.com/waku-org/specs/issues/38) to discuss this matter further. As decided on the research call (2024-09-09), these potential changes should be considered separately for a future iteration of the spec.

File: src/Membership.sol (hunk 6 above)
Comment (richard-ramos) [reply src/Membership.sol]: Billing period idea removed from code in https://github.com/waku-org/waku-rlnv2-contract/pull/13/commits/0cd80d639c1f6b3b0b8311dc6f1c1777e9a36323

File: src/Membership.sol (hunk 7 above)
Comment (richard-ramos) [reply src/Membership.sol]: Support for ETH removed in https://github.com/waku-org/waku-rlnv2-contract/pull/13/commits/7f60c538b320a50e12bf0c47fd3ccf0fc81e659e

File: src/Membership.sol (hunk 3 above)
Comment (richard-ramos) [reply src/Membership.sol:34]: Hm. I forgot that in the spec the max is set to 160_000 which exceeds the max of uint16

File: src/Membership.sol (hunk 9 above)
Comment (richard-ramos) [reply src/Membership.sol]: Removed the `TODO` from code in https://github.com/waku-org/waku-rlnv2-contract/commit/0cd80d639c1f6b3b0b8311dc6f1c1777e9a36323

File: src/Membership.sol (hunk 2 above)
Comment (richard-ramos) [reply src/Membership.sol]: Since the concept of billing periods was removed, these questions/comments are no longer relevant

File: src/Membership.sol (hunk 10 above)
Comment (richard-ramos) [reply src/Membership.sol]: @fryorcraken, @s-tikhomirov , 
In https://github.com/waku-org/waku-rlnv2-contract/pull/13/commits/8d43448de59a27e860576073753274f1ca6ecf65 the `register` function was modified and now exists as two  overloaded functions:
```
register(uint256 idCommitment, uint32 userMessageLimit);
register(uint256 idCommitment, uint32 userMessageLimit, uint256[] calldata membershipsToExpire)
```
I'm still working on the test units for the last function but this should cover what was discussed before.

Comment (fryorcraken) [reply src/Membership.sol]: > The idea of slot overwriting was initially meant to save contract storage space

Interesting, for me it was about booting off the network publishers with expired membership to keep network traffic and membership set closely related.

> every time, or is it perhaps better to do this (more) lazily?

As @richard-ramos ' stated:

```
register(uint256 idCommitment, uint32 userMessageLimit);
register(uint256 idCommitment, uint32 userMessageLimit, uint256[] calldata membershipsToErase)
```

With 2 functions on the contrat, it is now possible for the app to provide more feedback:

1. App sees there is space on the tree (`R_{max}` not reached yet, including new membership).
2. App can use `register(uint256 idCommitment, uint32 userMessageLimit);` and only an insertion happen

else

1. App sees there is **no** space on the tree (`R_{max}` is reached when including new membership).
2. App finds an `Expired` membership with greater or equal rate limit (a good actor would actually use the expired membership with greatest rate limit)
3. App can warn user that membership insertion would cost a bit more than usual
4. Insertion can be done with `register(uint256 idCommitment, uint32 userMessageLimit, uint256[] calldata membershipsToErase)`

This gives more predictability to the user.
App devs can also decide to spend the gas to purge `Expired` memberships to make it cheaper for users.

File: .gas-snapshot (hunk 11 above)
Comment (fryorcraken) [reply .gas-snapshot]: > We could probably cut back the cost by not mainainting the ordered membership list, and have nwaku be the one tracking the age of the memberships and deciding then which memberships to erase. I'd actually prefer that since we make the contract cheaper for the end user, but I don't know if keeping track of the membership age in nwaku is something we want to do?

do you mean that nwaku would need to listen to event to know age?

Is the issue that the list is ordered on the contract (I am not sure which piece of code you are referring with "due to")?

Why ordering the list?

File: src/Membership.sol (hunk 4 above)
Comment (alrevuelta) [reply src/Membership.sol:95]: index should be enough but not sure if the fact that memberships expire may require idcommit/msglimit. I mean, I don't think it would be strictly necessary but nice for checking that the index you are removing corresponds to the commitment you are expecting. But yeah, a nice to have, not required.

**Comment by jm-clius** [comment]
Great work!

@kaiserd @0x-r4bbit could we get a review from the Smart Contract team as well?

**Comment by 0x-r4bbit** [comment]
Looking into this now

**Review by 0x-r4bbit — COMMENTED** [review]
Left some comments.

Generally, you could make use of custom errors to save some gas

File: src/LinearPriceCalculator.sol [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+/// @title Linear Price Calculator to determine the price to acquire a membership
+contract LinearPriceCalculator is IPriceCalculator, Ownable {
+    /// @notice Address of the ERC20 token accepted by this contract. Address(0) represents ETH
+    address public token;
+
+    /// @notice The price per message per epoch
+    uint256 public pricePerMessagePerEpoch;
+
+    constructor(address _token, uint256 _pricePerMessagePerEpoch) Ownable() {
+        require(_token != address(0), "only tokens can be used");
```
[the remark this answers is not in the document]
Comment (richard-ramos) [reply src/LinearPriceCalculator.sol:16]: > In line: 9 we say that address(0) represents ETH as token.
Yet, here we're requiring _token != address(0).
Based on the comment, we either accept address(0) or adjust the comment.

I forgot to delete this when removing the support for ETH

File: src/LinearPriceCalculator.sol [hunk 13]
```diff
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+import { Ownable } from "openzeppelin-contracts/contracts/access/Ownable.sol";
+import { IPriceCalculator } from "./IPriceCalculator.sol";
+
+/// @title Linear Price Calculator to determine the price to acquire a membership
+contract LinearPriceCalculator is IPriceCalculator, Ownable {
```
[the remark this answers is not in the document]
Comment (richard-ramos) [reply src/LinearPriceCalculator.sol:8]: Sounds good! I'll apply this to the WakuRlnV2 contract as well!

File: src/WakuRlnV2.sol [hunk 14]
```diff
@@ -133,34 +135,65 @@ contract WakuRlnV2 is Initializable, OwnableUpgradeable, UUPSUpgradeable {
...
         external
         onlyValidIdCommitment(idCommitment)
-        onlyValidUserMessageLimit(userMessageLimit)
     {
-        _register(idCommitment, userMessageLimit);
+        if (memberExists(idCommitment)) revert DuplicateIdCommitment();
+
+        for (uint256 i = 0; i < membershipsToErase.length; i++) {
+            uint256 idCommitmentToErase = membershipsToErase[i];
+            MembershipInfo memory mdetails = members[idCommitmentToErase];
```
[the remark this answers is not in the document]
Comment (richard-ramos) [reply src/WakuRlnV2.sol:167]: > Otherwise it looks like this will be implicitly checked through things like grace period and expiry, correct?

Hm. Good point. It's probably better to be explicit in this case.

**Comment by richard-ramos** [comment]
Thank you for the review, @0x-r4bbit. I fixed the items and included your suggestion in the last commit

File: .gas-snapshot (hunk 11 above)
Comment (fryorcraken) [reply .gas-snapshot]: > The new membership SHOULD overwrite the membership that has been Expired for the longest time

I think it would make sense to drop this spec requirement and move to a model where the caller can pass a membership index to overwrite.

Ideally, we should store expiry time (or registration time + length) on the smart contract so we don't need to listen to events.

File: src/WakuRlnV2.sol [hunk 15]
```diff
@@ -133,34 +130,66 @@ contract WakuRlnV2 is Initializable, OwnableUpgradeable, UUPSUpgradeable {
         return rateCommitment != 0;
```
Comment (s-tikhomirov) [inline src/WakuRlnV2.sol:130]: It looks like this is the only usage of the function `idCommitmentToMetadata`, and we ignore two out of three values it returns. Would it be more efficient to implement something like `getRateCommitment` instead?

Comment (richard-ramos) [reply src/WakuRlnV2.sol:130]: Hmm indeed. We could probably just create a modifier that does `if (members[idCommitment].userMessageLimit != 0) revert DuplicateIdCommitment()` and use that in both `register` functions.

Comment (richard-ramos) [reply src/WakuRlnV2.sol:135]: Fixed in https://github.com/waku-org/waku-rlnv2-contract/pull/13/commits/522599c704702856e41a7abc5e2dba8ef2c0fb3d

File: src/WakuRlnV2.sol [hunk 16]
```diff
@@ -142,25 +144,61 @@ contract WakuRlnV2 is Initializable, OwnableUpgradeable, UUPSUpgradeable {
...
+    }
+
+    /// @dev Registers a member
+    /// @param idCommitment The idCommitment of the member
+    /// @param userMessageLimit The message limit of the member
+    /// @param index Indicates the index in the merkle tree
+    /// @param reusedIndex indicates whether we're inserting a new element in the merkle tree or updating a existing
+    /// leaf
+    function _register(uint256 idCommitment, uint32 userMessageLimit, uint32 index, bool reusedIndex) internal {
+        if (nextCommitmentIndex >= SET_SIZE) revert FullTree();
```
Comment (s-tikhomirov) [inline src/WakuRlnV2.sol:191]: Can this check be done in the two `register` functions alongside their mofidiers? That would allow us to save gas by reverting earlier if the membership set is full, before even calling `_register`.

File: src/WakuRlnV2.sol [hunk 17]
```diff
@@ -195,4 +233,75 @@ contract WakuRlnV2 is Initializable, OwnableUpgradeable, UUPSUpgradeable {
         }
         return castedProof;
     }
+
+    /// @notice Extend a membership expiration date. Memberships must be on grace period
+    /// @param idCommitments list of idcommitments
+    function extend(uint256[] calldata idCommitments) external {
+        for (uint256 i = 0; i < idCommitments.length; i++) {
+            uint256 idCommitment = idCommitments[i];
+            _extendMembership(_msgSender(), idCommitment);
```
Comment (s-tikhomirov) [inline src/WakuRlnV2.sol:242]: Should this be done in one line for brevity? 

```
_extendMembership(_msgSender(), idCommitments[i]);
```

**Review by jm-clius — COMMENTED** [review]
@richard-ramos @s-tikhomirov are there any blockers preventing this PR from being approved and merged? I know we may want to deploy this to a testnet to see how well it works in practice, but presumably we can address any new issues in subsequent PRs?

**Comment by richard-ramos** [comment]
@jm-clius
I havent merged it because of it not being approved :heavy_check_mark:  :(

**Comment by jm-clius** [comment]
@s-tikhomirov @0x-r4bbit could we get approvals if you're happy with scope of functionality on one hand and implementation on the other?

**Review by 0x-r4bbit — APPROVED** [review]
Sorry, didn't mean to block this here!

**Comment by richard-ramos** [comment]
Thank you, @0x-r4bbit and @s-tikhomirov!
